strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+
+    If Y is not specified, it defaults to 0.
+
+    Args:
+        covs (array): array of covariance matrices
+        X (array): matrix for mutltiplicative part of transformation
+        perm_out (array): indices for quadrature ordering
+        Y (array or 0): matrix for additive part of transformation.
+
+    Returns:
+        array: transformed array of covariance matrices
+    """
+    X_perm = X[:, perm_out][perm_out, :]
+    if not isinstance(Y, int):
+        Y = Y[:, perm_out][perm_out, :]
+    return (X_perm @ covs @ X_perm.T) + Y
+
+
+# pylint: disable=too-many-instance-attributes
+class BosonicModes:
+    """A Bosonic circuit class."""
+
+    # pylint: disable=too-many-public-methods
+
+    def __init__(self):
+        self.hbar = 2
+
+    def add_mode(self, peak_list=None):
+        r"""Add len(peak_list) modes to the circuit. Each mode has a number of
+        weights specified by peak_list, and the means and covariances are set
+        to vacuum.
+
+        Args:
+            peak_list (list): list of weights per mode.
+        """
+        if peak_list is None:
+            peak_list = [1]
+
+        num_modes = len(peak_list)
+        num_gauss = np.prod(peak_list)
+        self.nlen += num_modes
+
+        # Updated mode index permutation list
+        self.to_xp = to_xp(self.nlen)
+        self.from_xp = from_xp(self.nlen)
+        self.active.append(self.nlen - 1)
+
+        # Weights are set equal to each other and normalized
+        vac_weights = np.array([1 / num_gauss for i in range(num_gauss)], dtype=complex)
+        # New mode means and covs set to vacuum
+        vac_means = np.zeros((num_gauss, 2 * num_modes)).tolist()
+        vac_covs = [
+            ((self.hbar / 2) * np.identity(2 * num_modes)).tolist() for i in range(num_gauss)
+        ]
+
+        # Find all possible combinations of means and combs of the
+        # Gaussians between the modes.
+        mean_combs = it.product(self.means.tolist(), vac_means)
+        cov_combs = it.product(self.covs.tolist(), vac_covs)
+
+        # Tensor product of the weights.
+        weights = np.kron(self.weights, vac_weights)
+        # De-nest the means iterator.
+        means = np.array([[a for b in tup for a in b] for tup in mean_combs])
+        # Stack covs appropriately.
+        covs = np.array([block_diag(*tup) for tup in cov_combs])
+
+        self.weights = weights
+        self.means = means
+        self.covs = covs
+
+    def del_mode(self, modes):
+        r"""Delete modes from the circuit.
+
+        Args:
+            modes (int or list): modes to be deleted.
+        """
+        if isinstance(modes, int):
+            modes = [modes]
+
+        for mode in modes:
+            if self.active[mode] is None:
+                raise ValueError("Cannot delete mode, mode does not exist")
+
+            self.loss(0.0, mode)
+            self.active[mode] = None
+
+    def reset(self, num_subsystems=None, num_weights=1):
```

Comment:
Could there be cases where we would like to provision the weights ourselves, instead of initializing them to an equal superposition?

Reply:
The reset function is just to reset the modes to vacuum, but they're "nontraditional" vacuum states, in that they are linear combinations of Gaussians all centered at the origin of phase space with equal weight, such that when you add them up you get the vacuum. We do it this way because the other backends' reset functions revert the modes to vacuum, but we want the additional flexibility, through num_weights, to have arrays of means, weights and covs, that we can then manipulate. 

In general, we certainly want to initialize more complicated linear combinations of Gaussians, and we provide routines of interest in the next PR which includes state prep methods.